compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -675,4 +679,485 @@ object Erasure {
+        case PolyType(tparams, restpe) =>
+          assert(tparams.nonEmpty)
+          if (toplevel) polyParamSig(tparams)
+          builder.append("()")
+          methodResultSig(restpe)
+
+        case mtpe: MethodType =>
+          val params = mtpe.paramInfos
+          val restpe = mtpe.resultType
+          builder.append('(')
+          // TODO: Update once we support varargs
+          params.foreach { tp =>
+            jsig(tp)
+          }
+          builder.append(')')
+          methodResultSig(restpe)
+
+        case AndType(tp1, tp2) =>
+          jsig(intersectionDominator(tp1 :: tp2 :: Nil), primitiveOK = primitiveOK)
+
+        case ci: ClassInfo =>
+          def polyParamSig(tparams: List[TypeParamInfo]): Unit =
+            if (tparams.nonEmpty) {
+              builder.append('<')
+              tparams.foreach { tp =>
+                builder.append(sanitizeName(tp.paramName.lastPart))
+                boundsSig(hiBounds(tp.paramInfo.bounds))
+              }
+              builder.append('>')
+            }
+          val tParams = tp.typeParams
+          if (toplevel) polyParamSig(tParams)
+          superSig(ci.typeSymbol, ci.parents)
+
+        case AnnotatedType(atp, _) =>
+          jsig(atp, existentiallyBound, toplevel, primitiveOK)
+
+        case hktl: HKTypeLambda =>
+          jsig(hktl.finalResultType)
+
+        case _ =>
+          val etp = erasure(tp)
+          if (etp eq tp) throw new UnknownSig
+          else jsig(etp)
+      }
+    }
+    val throwsArgs = sym0.annotations flatMap ThrownException.unapply
+    if (needsJavaSig(info, throwsArgs)) {
+      try {
+        jsig(info, toplevel = true)
+        throwsArgs.foreach { t =>
+          builder.append('^')
+          jsig(t, toplevel = true)
+        }
+        Some(builder.toString)
+      }
+      catch { case _: UnknownSig => None }
+    }
+    else None
+  }
+
+  // TODO: Port from scalac
+  private object ThrownException {
+    def unapply(ann: Annotation): Option[Type] = None
+  }
+
+  class UnknownSig extends Exception
+
+  private def needsJavaSig(tp: Type, throwsArgs: List[Type])(implicit ctx: Context): Boolean = !ctx.settings.YnoGenericSig.value && {
+      def needs(tp: Type) = (new NeedsSigCollector).apply(false, tp)
+      needs(tp) || throwsArgs.exists(needs)
+  }
+
+  // @M #2585 when generating a java generic signature that includes
+  // a selection of an inner class p.I, (p = `pre`, I = `cls`) must
+  // rewrite to p'.I, where p' refers to the class that directly defines
+  // the nested class I.
+  //
+  // See also #2585 marker in javaSig: there, type arguments must be
+  // included (use pre.baseType(cls.owner)).
+  //
+  // This requires that cls.isClass.
+  private def rebindInnerClass(pre: Type, cls: Symbol)(implicit ctx: Context): Type = {
+    val owner = cls.owner
+    if (owner.is(Flags.PackageClass) || owner.isTerm) pre else cls.owner.info /* .tpe_* */
+  }
+
+  object GenericArray {
```

Comment:
It would be good to base this on the `Definitions.MultiArrayOf` extractor. This would avoid duplication and simplify the code that needs to be added.

Reply:
I simplified `GenericArray` as much as I could, but I cannot use `MultiArrayOf` directly, since it'll resolve the upper bound when encountering type bounds. This means that types such as
```scala
class Foo[T <: Array[Int]]
```
would get a wrong signature (`T <: java.lang.Object` instead of `T <: int[]`)